pkg/provider/ibmcloud/plugin/loadbalancer/lb.go

```diff
@@ -0,0 +1,488 @@
+func max(x, y int) int {
+	if x > y {
+		return x
+	}
+	return y
+}
+
+// Name is the name of the load balancer
+func (l *ibmcloudlb) Name() string {
+	lbLogger.Debug("Name", "V", debugV)
+	return l.name
+}
+
+// Routes lists all known routes.
+func (l *ibmcloudlb) Routes() ([]loadbalancer.Route, error) {
+	lbLogger.Debug("Routes", "V", debugV)
+	l.lock.Lock()
+	defer l.lock.Unlock()
+
+	// Output array
+	out := []loadbalancer.Route{}
+
+	// Get the listeners for the loadbalancer
+	lb, err := l.lbService.Mask("listeners;listeners.defaultPool").GetLoadBalancer(&l.uuid)
+	if err != nil {
+		return out, err
+	}
+
+	// Iterate over the listeners filling the output array
+	for _, listener := range lb.Listeners {
+		var certString string
+		if listener.TlsCertificateId != nil {
+			cert, err := l.certService.Id(*listener.TlsCertificateId).GetObject()
+			if err != nil {
+				return out, err
+			}
+			certString = fmt.Sprintf("%s(%s)", *cert.CommonName, *cert.Notes)
+		}
+		r := loadbalancer.Route{
+			LoadBalancerPort:     *listener.ProtocolPort,
+			LoadBalancerProtocol: loadbalancer.ProtocolFromString(*listener.Protocol),
+			Port:                 *listener.DefaultPool.ProtocolPort,
+			Protocol:             loadbalancer.ProtocolFromString(*listener.DefaultPool.Protocol),
+		}
+		if certString != "" {
+			r.Certificate = &certString
+		}
+		out = append(out, r)
+	}
+
+	lbLogger.Debug("Routes", "name", l.name, "V", debugV, "routes", out)
+	return out, nil
+}
+
+// Publish publishes a route in the LB by adding a load balancing rule
+func (l *ibmcloudlb) Publish(route loadbalancer.Route) (loadbalancer.Result, error) {
+	lbLogger.Debug("Publish", "name", l.name, "route", route, "V", debugV)
+	l.lock.Lock()
+	defer l.lock.Unlock()
+
+	// Get the listeners for the loadbalancer
+	lb, err := l.lbService.Mask("listeners").GetLoadBalancer(&l.uuid)
+	if err != nil {
+		return nil, err
+	}
+
+	// Iterate over the listeners to see is the specified port is already in use.
+	// Log Info message if there is.
+	for _, listener := range lb.Listeners {
+		if *listener.ProtocolPort == route.LoadBalancerPort {
+			// return lbResult(""), fmt.Errorf("duplicate port %v", route.LoadBalancerPort)
+			lbLogger.Error("Publish", "name", l.name, "route", route, "V", debugV, "Duplicate Port", route.LoadBalancerPort)
+			return lbResult("publish"), nil
+		}
+	}
+
+	// Get certificate id if certificate name is provided
+	var certID int
+	if route.Certificate != nil && *route.Certificate != "" {
+		// The certificate string is the common name of the certificate.
+		cert, err := l.certService.FindByCommonName(route.Certificate)
+		if err != nil {
+			return nil, err
+		}
+		if len(cert) == 1 {
+			certID = *cert[0].Id
+		} else {
+			return nil, fmt.Errorf("Cannot identify an unique certificate with common name '%s'", *route.Certificate)
+		}
+	}
+
+	// InfraKit doesn't define a method, so default to round robin
+	lbMethod := "ROUNDROBIN"
```

Comment:
Is this something that we can add to the `Route` struct?  Different L4 implementations will have different strings but if this is just some opaque value that's passed on to the L4 then we can avoid this hardcoding here.

Reply:
The specification of the method could be added. For information, the IBM LBaaS LB supports 3 methods: Round Robin, Weighted Round Robin (would also require the addition of a weight on each backend), and Least Connection.